### PR TITLE
88 run the services and factories

### PR DESCRIPTION
**Pull request by Liranitz**

### CODE REVIEW

File: Market-Trading-System/src/main/java/DomainLayer/Market/Main.java [hunk 1]
```diff
@@ -1,8 +1,30 @@
...
 public class Main {
 
     public static void main(String[] args) {
-        System.out.println("Hello world!");
+        // Automatically create a SystemManager instance with relevant data from the User constructor
+        SystemManager systemManager = SystemManager.getInstance();
+
+        // Command-line interface
+        Scanner scanner = new Scanner(System.in);
+        System.out.println("Hey Ish Yakar, The system is empty please enter as admin, Welcome to the Market System Manager CLI!");
```
Comment (NoamShurki) [inline Market-Trading-System/src/main/java/DomainLayer/Market/Main.java]: change the context of the text

File: Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/OutServices/PaymentServiceImpl.java [hunk 2]
```diff
@@ -1,10 +1,20 @@
...
 
 import DomainLayer.Market.Purchase.Abstractions.IPaymentService;
+import ServiceLayer.ServiceFactory;
 
 import java.util.Date;
 
 public class PaymentServiceImpl implements IPaymentService {
+    private static PaymentServiceImpl instance;
+
+    public static PaymentServiceImpl getInstance() {
```
Comment (NoamShurki) [inline Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/OutServices/PaymentServiceImpl.java]: add private constructor

File: Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/PaymentServiceProxy.java [hunk 3]
```diff
@@ -8,8 +8,16 @@
...
     private PaymentServiceImpl paymentService;
 
+    public static PaymentServiceProxy getInstance(PaymentServiceImpl paymentService) {
+        if(instance == null) {
+            instance = new PaymentServiceProxy(paymentService);
+        }
+        return instance;
+    }
+
     public PaymentServiceProxy(PaymentServiceImpl paymentService) {
```
Comment (NoamShurki) [inline Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/PaymentServiceProxy.java]: change constructor to private

File: Market-Trading-System/src/main/java/DomainLayer/Market/User/IUserFacade.java [hunk 4]
```diff
@@ -5,8 +5,19 @@
...
+import DomainLayer.Market.Util.IRepository;
+
 public interface IUserFacade {
-    public String createGuestSession();
+    public static IUserFacade getInstance(IRepository<String, User> users)
+    {
+        return UserController.getInstance(users);
+    }
+    public void setStoreFacade(IStoreFacade storeFacade);
+    public void createGuestSession();
```
Comment (NoamShurki) [inline Market-Trading-System/src/main/java/DomainLayer/Market/User/IUserFacade.java]: change return type back to string

File: Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/SupplyServiceProxy.java [hunk 5]
```diff
@@ -7,12 +7,20 @@
...
 public class SupplyServiceProxy implements ISupplyService {
+    private static SupplyServiceProxy instance;
     private SupplyServiceImpl supplyService;
 
-    public SupplyServiceProxy(SupplyServiceImpl supplyService) {
+    private SupplyServiceProxy(SupplyServiceImpl supplyService) {
         this.supplyService = supplyService;
     }
 
+    public static SupplyServiceProxy getInstance(SupplyServiceImpl supplyServiceInstance) {
```
Comment (NoamShurki) [inline Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/SupplyServiceProxy.java]: add synchronize

File: Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/PurchaseController.java [hunk 6]
```diff
@@ -17,44 +17,39 @@
...
-        this.supplyServiceImpl=supplyServiceImpl;
+    private PurchaseController(IRepository<Long,Purchase> purchaseRepo,PaymentServiceProxy paymentServiceProxy, SupplyServiceProxy supplyServiceProxy) {
+        this.purchaseRepo = purchaseRepo;
+        this.paymentServiceProxy = paymentServiceProxy;
+        this.supplyServiceProxy = supplyServiceProxy;
 
         inventoryReduceItems = new PriorityBlockingQueue<ItemDTO>(); //protected queue
+    }
 
+    public static PurchaseController getInstance(IRepository<Long, Purchase> purchaseRepo, PaymentServiceProxy paymentServiceProxy, SupplyServiceProxy supplyServiceProxy) {
```
Comment (NoamShurki) [inline Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/PurchaseController.java]: add synchronized

File: Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/PaymentServiceProxy.java [hunk 7]
```diff
@@ -8,8 +8,16 @@
 import java.util.List;
 
 public class PaymentServiceProxy implements IPaymentService {
+    private static PaymentServiceProxy instance;
     private PaymentServiceImpl paymentService;
 
+    public static PaymentServiceProxy getInstance(PaymentServiceImpl paymentService) {
```
Comment (NoamShurki) [inline Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/PaymentServiceProxy.java]: add shyncronize

File: Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/OutServices/SupplyServiceImpl.java [hunk 8]
```diff
@@ -1,8 +1,20 @@
...
 
 import DomainLayer.Market.Purchase.Abstractions.ISupplyService;
+import DomainLayer.Market.Purchase.SupplyServiceProxy;
 
 public class SupplyServiceImpl implements ISupplyService {
+    private static SupplyServiceImpl instance;
+
+    private SupplyServiceImpl() {}
+
+    public static SupplyServiceImpl getInstance() {
```
Comment (NoamShurki) [inline Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/OutServices/SupplyServiceImpl.java]: add synchronize

File: Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/OutServices/PaymentServiceImpl.java (hunk 2 above)
Comment (NoamShurki) [inline Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/OutServices/PaymentServiceImpl.java]: add syncronized

File: Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/IPurchaseFacade.java [hunk 9]
```diff
@@ -4,12 +4,17 @@
...
 import DomainLayer.Market.Purchase.OutServices.SupplyServiceImpl;
 import DomainLayer.Market.Store.Item;
+import DomainLayer.Market.Util.IRepository;
 
 import java.util.Date;
 import java.util.HashMap;
 import java.util.List;
 
 public interface IPurchaseFacade {
+    public static IPurchaseFacade getInstance(IRepository<Long, Purchase> purchaseRepo, PaymentServiceProxy paymentServiceProxy, SupplyServiceProxy supplyServiceProxy)
```
Comment (NoamShurki) [inline Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/IPurchaseFacade.java:14]: add syncronize

**Review by NoamShurki — CHANGES_REQUESTED** [review]

File: Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/IPurchaseFacade.java (hunk 9 above)
Comment (Liranitz) [reply Market-Trading-System/src/main/java/DomainLayer/Market/Purchase/IPurchaseFacade.java:14]: no need in facade

**Review by NoamShurki — APPROVED** [review]

**Pull request merged**